### ISSUE TITLE
Detect the number of affected rows by INSERT/UPDATE/DELETE.

### ISSUE THREAD

**Issue by MaslovGennady** [ClickHouse/clickhouse-odbc#335]
As mentioned in 

https://github.com/ClickHouse/clickhouse-odbc/issues/324#issuecomment-724233892

comment we need the clickhouse ODBC driver to be able to detect the number of affected rows by INSERT/UPDATE/DELETE (to return correct result for SQLRowCount function).
This is for developing ETL on informatica powercenter that connects to clickhouse with ODBC.
Without this feature Informatica gets 0 as return value from SQLRowCount function when inserting data to clickhouse through ODBC. Informatica then decides that 0 rows have been inserted and repeats insert. So this inserting repeats infinitely.

**Comment by singhpratech** [comment]
Still the case on 1.5.5.20260810 against 26.7.5.10 over HTTP: `SQLRowCount` returns `SQL_SUCCESS` with the value **0** after a single-row `INSERT`, a 3-row `INSERT ... VALUES (...),(...),(...)`, a prepared parameterised `INSERT`, an `ALTER TABLE ... DELETE` and a `CREATE TABLE` alike — checked with a plain-ODBC C program (unixODBC 2.3.12, Linux), `Memory` and `MergeTree`. Worth noting for anyone reading this: 0 is a stated count, not the ODBC "unknown" value −1, so a generic caller that trusts the count concludes nothing was written. If the count is not knowable from the HTTP response, −1 would at least be honest — but it is knowable: the server sends `X-ClickHouse-Summary: {"read_rows":"3",...,"written_rows":"3",...}` on every HTTP response by default (checked with a plain `curl` `INSERT ... VALUES (1),(2),(3)` against 26.7.5.10), so the driver could take `written_rows` from the header it already receives.

**Comment by slabko** [comment]
@singhpratech, returning `0` by default is indeed incorrect behavior. This absolutely needs to be fixed. However, I'm not sure with X-ClickHouse-Summary. This values provided by this header are not very reliable, see https://github.com/ClickHouse/ClickHouse/issues/57768. It should be possible to use this header when `async_insert` is disabled, but I still need to think about how to handle this correctly. For now, I will fix the driver so that it returns `-1` instead of `0`. I hope that will unblock you.

Thank you for the reminder. This issue indeed fell under my radar.

**Comment by singhpratech** [comment]
Thank you, that unblocks the honest case straight away: −1 is what ODBC reserves for "not known", and a caller that reads it treats the count as unavailable instead of as zero rows written. On the [adbcBridge](https://github.com/singhpratech/adbcbridge) side that means `rows_affected` comes back as −1 on ClickHouse, which is what the ADBC contract allows for a driver that cannot count, and the compatibility matrix will stop carrying a tolerance for the zero.

On `X-ClickHouse-Summary`: agreed that with `async_insert` on, the number the header reports is not the rows the statement itself wrote, so it cannot be trusted unconditionally. For the synchronous case it matched in every check I ran (single-row, multi-row `VALUES`, prepared parameterised inserts, a mutation), so a later step could be to use `written_rows` only when the session has `async_insert` off and keep −1 otherwise, which stays honest either way. Happy to run the matrix against a build of the −1 change whenever it is up.

**Comment by slabko** [comment]
Just realized that `written_rows` in `X-ClickHouse-Summary` also reports rows changed in the associated materialized views.

**Comment by singhpratech** [comment]
You are right, and thank you for adding that: if the number can include the materialized views, it is not the affected-row count for the target table, and returning it as one would only trade a wrong 0 for a wrong 7. -1 is the honest answer for the ODBC contract, and leaving the header alone seems like the right call to me as well.

In case it is useful to have on record, this is what I measured on 26.7.5.10 with two materialized views on the target (one plain, one aggregating), reading `X-ClickHouse-Summary` with `wait_end_of_query=1` and cross-checking `system.query_log`:

| insert of 3 rows | `written_rows` |
|---|---|
| `INSERT ... VALUES`, no views | 3 |
| `INSERT ... VALUES`, two views | 3 |
| `INSERT ... SELECT`, two views | 7 (3 target + 3 plain view + 1 aggregate row) |

So at least on this version the views are counted on the `INSERT ... SELECT` path and not on the `VALUES` path, which I read as one more reason not to treat the header as a per-table count. If you ever see a way to get a per-table number from the server, I would be glad to test it from the adbcBridge side (https://github.com/singhpratech/adbcbridge); until then -1 is exactly what it passes through, and that is a much better place than where this started. Thank you again.
